partition/gpt/table.go

```diff
@@ -396,22 +390,88 @@ func tableFromBytes(b []byte, logicalBlockSize, physicalBlockSize int) (*Table,
 		return nil, fmt.Errorf("invalid EFI Header Checksum, expected %v, got %v", checksum, efiHeaderCrc)
 	}
 
-	// potential protective MBR is at LBA0
-	hasProtectiveMBR := readProtectiveMBR(b[:logicalBlockSize], uint32(secondaryHeader))
-
 	table := Table{
-		LogicalSectorSize:      logicalBlockSize,
-		PhysicalSectorSize:     physicalBlockSize,
 		partitionEntrySize:     partitionEntrySize,
 		primaryHeader:          primaryHeader,
 		secondaryHeader:        secondaryHeader,
 		firstDataSector:        firstDataSector,
 		lastDataSector:         lastDataSector,
 		partitionArraySize:     int(partitionEntryCount),
 		partitionFirstLBA:      partitionEntryFirstLBA,
-		ProtectiveMBR:          hasProtectiveMBR,
 		GUID:                   strings.ToUpper(diskGUID.String()),
 		partitionEntryChecksum: partitionEntryChecksum,
+	}
+
+	return &table, nil
+}
+
+// tableHeaderFromBytes read a partition table from a byte slice, mainly used to validate the secondary header
+func tableHeaderFromBytes(b []byte, logicalBlockSize, physicalBlockSize int, skipMBR bool) (*Table, error) {
+	// minimum size - gpt entries + header + LBA0 for (protective) MBR
+	minSize := logicalBlockSize
+	if len(b) < minSize {
+		return nil, fmt.Errorf("data for partition was %d bytes instead of expected minimum %d", len(b), minSize)
+	}
+	gpt := b
+	if skipMBR {
+		gpt = b[logicalBlockSize:]
+	}
+
+	header, err := readGptHeader(gpt)
+	if err != nil {
+		return nil, err
+	}
+
+	// potential protective MBR is at LBA0
+	hasProtectiveMBR := readProtectiveMBR(b[:logicalBlockSize], uint32(header.secondaryHeader))
+
+	table := Table{
+		LogicalSectorSize:  logicalBlockSize,
+		PhysicalSectorSize: physicalBlockSize,
+		partitionEntrySize: header.partitionEntrySize,
+		primaryHeader:      header.primaryHeader,
+		secondaryHeader:    header.secondaryHeader,
+		firstDataSector:    header.firstDataSector,
+		lastDataSector:     header.lastDataSector,
+		partitionArraySize: header.partitionArraySize,
+		ProtectiveMBR:      hasProtectiveMBR,
+		GUID:               header.GUID,
+		initialized:        true,
+	}
+	return &table, nil
+}
+
+// tableFromBytes read a partition table from a byte slice
+func tableFromBytes(b []byte, logicalBlockSize, physicalBlockSize int) (*Table, error) {
+	// minimum size - gpt entries + header + LBA0 for (protective) MBR
+	if len(b) < logicalBlockSize*2 {
+		return nil, fmt.Errorf("data for partition was %d bytes instead of expected minimum %d", len(b), logicalBlockSize*2)
+	}
+
+	// GPT starts at LBA1
+	gpt := b[logicalBlockSize:]
+
+	header, err := readGptHeader(gpt)
+	if err != nil {
+		return nil, err
+	}
+
+	// potential protective MBR is at LBA0
+	hasProtectiveMBR := readProtectiveMBR(b[:logicalBlockSize], uint32(header.secondaryHeader))
+
+	table := Table{
```

Comment:
Rather than copying the fields, since `readGptHeader()` returns `*Table`, wouldn't it be easier to just add the data to it and return it? Something like:

```go
table, err := readGptHeader(gpt)
// .. lots of other stuff
table.LogicalSectorSize = logicalBlockSize
table.PhysicalSectorSize = physicalBlockSize
table.ProtectiveMBR = hasProtectiveMBR
table.initialized = true
```